sky/task.py

```diff
@@ -621,18 +622,38 @@ def set_resources_override(self, override_params: Dict[str, Any]) -> 'Task':
     def service(self) -> Optional[service_spec.SkyServiceSpec]:
         return self._service
 
-    def set_service(self,
-                    service: Optional[service_spec.SkyServiceSpec]) -> 'Task':
+    def set_service(self, service: Optional[service_spec.SkyServiceSpec]):
         """Sets the service spec for this task.
-
         Args:
           service: a SkyServiceSpec object.
-
-        Returns:
-          self: The current task, with service set.
         """
+
+        if service is None:
+            return
+
+        if service.spot_zones is None and service.spot_placer is not None:
+
+            logger.info('service.spot_zones is not specified, '
+                        'use all available zones')
+
+            launchable_resources, _, _ = (
+                optimizer.fill_in_launchable_resources(
+                    task=self,
+                    blocked_resources=None,
+                    try_fix_with_sky_check=True,
+                    quiet=True))
+
+            launchable_zones = set()
+            for res in launchable_resources[list(self.resources)[0]]:
+                regions = res.get_valid_regions_for_launchable()
+                for region in regions:
+                    if region.zones is not None:
+                        for zone in region.zones:
+                            launchable_zones.add(zone.name)
```

Comment:
Feeling like it should be done inside the service_spec, rather than `sky.Task`.

Reply:
It requires calling `optimizer.fill_in_launchable_resources` on `task`. There could be other user-specified task constraints in the yaml. I don't think we should instantiate task in service_spec.